text/0812-tracked-built-ins.md

```diff
@@ -0,0 +1,287 @@
+in the reactivity mental model we have today when it comes to tracking updates in data structures.
+
+Today, `@tracked` decorator may be used to track updates of the primitive values:
+
+```js
+import Component from '@glimmer/component';
+import { tracked } from '@glimmer/tracking';
+import { action } from '@ember/object';
+
+export default class CounterComponent extends Component {
+  @tracked count = 0;
+
+  @action increment() {
+    this.count = this.count + 1;
+  }
+
+  @action decrement() {
+    this.count = this.count - 1;
+  }
+}
+```
+
+However, there are cases when there are unknown, dynamic, or potentially infinite numbers
+of decoratable keys. In such cases we can use `Object`, `Map` or another data structure
+that fits most in the particular use case. The initial pattern proposed for Ember Octane
+in such scenarios was to re-set the tracked property:
+
+```js
+class TrackedMap {
+  @tracked map = new Map();
+
+  updateMapValue(key, value) {
+    this.map.set(key, value);
+    this.map = this.map;
+  }
+}
+```
+
+This pattern is problematic however, because it's not obvious what the
+purpose of re-setting the property is. A developer unfamiliar with Ember may
+mistakenly believe this was an error, and remove this statement, causing
+bugs. It's not a very google-able pattern either, which makes it more
+difficult to learn and to teach. Finally, it's fairly easy this way for the
+state to get out of sync with the rest of the system if a developer forgets
+to add the extra re-set after mutating the value.
+
+Over time, it's become clear that this pattern is in actuality an antipattern
+for these reasons, and something we should begin to discourage in general.
+
+[RFC #669 "Tracked Storage Primitives"][rfc-0669] introduced low level primitives
+that allow to implement tracked versions of JavaScript's built-in classes
+in an addon like [tracked-built-ins](https://github.com/tracked-tools/tracked-built-ins):
+
+```ts
+import {
+  TrackedObject,
+  TrackedArray,
+  TrackedMap,
+  TrackedSet,
+  TrackedWeakMap,
+  TrackedWeakSet,
+} from 'tracked-built-ins';
+
+class Foo {
+  @tracked value = 123;
+
+  obj = new TrackedObject<{[string: number]}>({ foo: 123 });
+  arr = new TrackedArray<string, number>(['foo', 123]);
+  map = new TrackedMap<string, number>(['foo', 123]);
+  set = new TrackedSet<string, number>(['foo', 123]);
+  weakMap = new TrackedWeakMap<{[object: number]}>([[{}, 456]]);
+  weakSet = new TrackedWeakSet<string, number>(['foo', 123]);
+}
+```
+
+or using `tracked()` decorator:
+
+```ts
+import { tracked } from 'tracked-built-ins';
+
+class Foo {
+  @tracked value = 123;
+
+  obj = tracked<{[string: number]}>({ foo: 123 });
+  arr = tracked<string, number>(['foo', 123]);
+  map = tracked<string, number>(new Map(['foo', 123]));
+  set = tracked<string, number>(new Set(['foo', 123]));
+  weakMap = tracked<{[string: number]}>(new WeakMap([[{}, 456]]));
+  weakSet = tracked<string, number>(new WeakSet(['foo', 123]));
+}
+```
```

Comment:
Raising this explicitly for discussion: I do not think we should include this part of the `tracked-built-ins` API.

Reply:
I agree with this. In practice, I've found that importing the decorator from `tracked-built-ins` gets confusing. In some components, you'll import it from `@glimmer/tracking`, and in others you might use `tracked-built-ins` instead. I find it much clearer to import the classes and instantiate them instead. We've even found rare cases where we need to use both:

```
@tracked someMap = new TrackedMap();
```

Being able to consistently understand what `@tracked` is capable of no matter which file you're looking at without having to check where it's imported from is much better for developer ergonomics, IMO.

Reply:
I tend to agree with this - `tracked()` function may be *misleading* and you need to be careful what you use and where.

would be puzzle to have in the file both `tracked` provided by `@glimmer/tracking` *and* `tracked` provided by `tracked-built-ins` as you would have to rename one of those in the import statement